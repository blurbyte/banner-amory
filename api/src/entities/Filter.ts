import { Item, Tier } from './Item';

// 0 means ALL parts and ranks
export type Filter = {
  part: 0 | Item['gamePart'];
  rank: 0 | Item['rank'];
  bonus: FilterBonus;
  tier: FilterTier;
};

export enum FilterBonus {
  Any = 'ANY',
  MainStat = 'MAIN_STAT',
  Talent = 'TALENT',
  PerTurn = 'PER_TURN',
  OnRest = 'ON_REST',
  Chance = 'CHANCE',
  FromMarket = 'FROM_MARKET',
  Movement = 'MOVEMENT'
}

export type FilterTier = 'ANY' | Tier;

// Item properties on which filtering is based
export type AvailableFilterBonusProperty = Extract<
  keyof Item,
  | 'statStrength'
  | 'statWillpower'
  | 'statArmor'
  | 'statArmorBreak'
  | 'talentAllRanks'
  | 'talentStrength'
  | 'talentWillpower'
  | 'talentArmor'
  | 'talentArmorBreak'
  | 'talentExertion'
  | 'perTurnArmor'
  | 'perTurnWillpower'
  | 'onRestArmor'
  | 'onRestWillpower'
  | 'chanceCrit'
  | 'chanceDodge'
  | 'chanceDivert'
  | 'acquirementMarketplace'
  | 'movement'
>;
