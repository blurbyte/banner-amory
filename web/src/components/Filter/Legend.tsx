import styled from 'styled-components';
import BaseLabel from '../Label';

const Paragraph = BaseLabel.withComponent('p');

const Legend = styled(Paragraph)`
  margin: 0;
  margin-right: 1rem;
  line-height: 2.4rem;
`;

export default Legend;
