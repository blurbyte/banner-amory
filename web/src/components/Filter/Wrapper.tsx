import styled from 'styled-components';

const Wrapper = styled.div`
  display: flex;
  margin-right: 2rem;
  margin-top: 1rem;
`;

export default Wrapper;
