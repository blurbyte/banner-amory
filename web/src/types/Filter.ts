// for sake of simplicity input values are strings and compability with GraphQL

export type Filter = {
  part: 'all' | '1' | '2';
  rank: 'all' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10';
  bonus: FilterBonus;
};

export enum FilterBonus {
  Any = 'ANY',
  MainStat = 'MAIN_STAT',
  Talent = 'TALENT',
  PerTurn = 'PEWR_TURN',
  OnRest = 'ON_REST',
  Chance = 'CHANCE',
  FromMarket = 'FROM_MARKET',
  Movement = 'MOVEMENT'
}
