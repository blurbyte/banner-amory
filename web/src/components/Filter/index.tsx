// group of filter items
// with automatic name, onChange and checkedValue applied to each child

import * as React from 'react';

import { Filter as FilterType } from '@sharedTypes/Filter';
import FilterItem from '../FilterItem';
import Wrapper from './Wrapper';
import Legend from './Legend';

export type FilterProps = {
  children: React.ReactNode;
  name: keyof FilterType;
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  checkedValues: FilterType;
  label?: string;
};

class Filter extends React.Component<FilterProps> {
  static Item = FilterItem;

  renderChildren = () => {
    const { children, name, onChange, checkedValues } = this.props;
    let checkedValue = checkedValues[name];

    // pass name prop to all Item children
    return React.Children.map(children, child => {
      return React.cloneElement(child as React.ReactElement<any>, { name, onChange, checkedValue });
    });
  };

  render() {
    const { name, label } = this.props;

    return (
      <Wrapper>
        <Legend>{label ? label : name}</Legend>
        {this.renderChildren()}
      </Wrapper>
    );
  }
}

export default Filter;
