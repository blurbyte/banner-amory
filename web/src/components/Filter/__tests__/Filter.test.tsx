import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { Filter as FilterType, FilterBonus } from '@sharedTypes/Filter';
import Filter from '../index';

test('renders correctly', () => {
  const checkedValues: FilterType = {
    part: 0,
    rank: 0,
    bonus: FilterBonus.Any
  };

  const component = renderer.create(
    <Filter name="rank" checkedValues={checkedValues} onChange={jest.fn()}>
      <Filter.Item value={0} />
      <Filter.Item value={1} />
    </Filter>
  );
  expect(component.toJSON()).toMatchSnapshot();
});
