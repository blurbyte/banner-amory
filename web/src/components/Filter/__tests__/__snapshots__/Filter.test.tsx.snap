// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="sc-htpNat jeiAyX"
>
  <p
    className="sc-EHOje euXMrm sc-ifAKCX bVJDvB"
  >
    rank
  </p>
  <label
    checked={true}
    className="sc-bdVaJa gsvaU"
  >
    0
    <input
      checked={true}
      className="sc-bwzfXH kxRMuH"
      name="rank"
      onChange={[MockFunction]}
      type="radio"
      value={0}
    />
  </label>
  <label
    checked={false}
    className="sc-bdVaJa bJrhxk"
  >
    1
    <input
      checked={false}
      className="sc-bwzfXH kxRMuH"
      name="rank"
      onChange={[MockFunction]}
      type="radio"
      value={1}
    />
  </label>
</div>
`;
